Drivers/Periphs/Src/IMU.c

```diff
@@ -0,0 +1,290 @@
+
+    // TODO: move this into a function that checks if we've uncalibrated. 
+    uint8_t calib_reg = 0x00; 
+    while((calib_reg >> 6) != 3){ //3C is gyro, accel ;;FF is sys, gyr, accel, magnet ;;x30 is just gyro
+        HAL_I2C_Mem_Read(&hi2c1, ADDR, CALIB_STAT, I2C_MEMADD_SIZE_8BIT, &calib_reg, 1, HAL_MAX_DELAY);
+    }
+        
+    return error;
+}
+
+/**
+ * @brief Update struct with new information
+ * @param *Data : struct used to collect IMU Data
+ * @return HAL_StatusTypeDef - OK, ERROR, BUSY, or TIMEOUT
+ */
+HAL_StatusTypeDef IMU_GetMeasurements(IMUData_t *Data){
+    // Read 18 contiguous data registers within the IMU starting at 0x08 and stores data within struct fields
+    return HAL_I2C_Mem_Read(&hi2c1, ADDR, ACC_DATA_X_LSB, I2C_MEMADD_SIZE_8BIT, (uint8_t*)Data, 18, HAL_MAX_DELAY);
+}
+
+/**
+ * @brief Update struct with IMU Calibration register data. Used to hardcode calibration
+ * @note The IMU must be in config mode (REG[0x3D]= 0) to read calibration registers
+ * @param *Data : struct used to collect IMU calibration data
+ * @return HAL_StatusTypeDef - OK, ERROR, BUSY, or TIMEOUT
+ */
+HAL_StatusTypeDef IMU_GetCalibData(IMUCalibData_t *Data){
+
+    HAL_StatusTypeDef error = HAL_OK;
+    // This function has to execute while the IMU is in configuration mode. Otherwise the registers cannot be read
+    // TODO: check if we need to be in config mode to read these registers
+    
+    uint8_t config[4];
+    config[0] = OPR_MODE; //register address
+    config[1] = 0; // set to some configuration mode
+    error |= SEND(config, 2); // set IMU to configuration mode to extract calibration data
+    
+   //Read 22 contiguous bytes of calibration registers in the IMU. Load into calibration data struct fields
+    error |= HAL_I2C_Mem_Read(&hi2c1, ADDR, ACC_OFFSET_X_LSB, I2C_MEMADD_SIZE_8BIT, (uint8_t*)Data, 22, HAL_MAX_DELAY);
+    return error;
+}
+
+/**
+ * @brief Use IMUCalibData_t struct to upload calibration profile to IMU
+ * @note updating the last bytes of the Z offsets will set the respective peripheral bits in the calibration status register (0x35)
+ * @note The IMU must be in config mode (REG[0x3D]= 0) to read calibration registers
+ * @param *Data : struct holding IMU calibration data
+ * @return HAL_StatusTypeDef - OK, ERROR, BUSY, or TIMEOUT
+ */
+HAL_StatusTypeDef IMU_Calibrate(IMUCalibData_t *Data){
+
+    HAL_StatusTypeDef error = HAL_OK;
+    // This function must be executed while the IMU is in configuration mode. Otherwise registers are not writable
+    uint8_t config[4];
+    config[0] = OPR_MODE; //register address
+    config[1] = 0; // set to some configuration mode
+    error |= SEND(config, 2); // set IMU to configuration mode to extract calibration data
+    
+    int16_t index =0x55;
+    for (int32_t reg=0; reg < 22; reg +=1) {
+        config[0] = index;
+        config[1] = 0x01;
+        error |=SEND(config, 2);
+        index +=1;
+    }
+
+    index =0x67;
```

Comment:
if you end up using `index` (the `index` variable isn't really necessary in this function), make its type a `uint8_t` since `*config` is getting set to `index`.